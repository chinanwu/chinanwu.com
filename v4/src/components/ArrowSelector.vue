<template>
  <div class="ArrowSelector">
    <div class="ArrowSelector__main">
      <button
        id="arrowSelectorLeft"
        class="ArrowSelector__btn"
        @click="handleLeftArrow"
        aria-label="Select previous project"
      >
        &lt;
      </button>
      <span aria-labelledby="arrowSelectorLabel"
        >{{ current }} / {{ total }}</span
      >
      <button
        id="arrowSelectorRight"
        class="ArrowSelector__btn"
        @click="handleRightArrow"
        aria-label="Select next project"
      >
        &gt;
      </button>
    </div>
    <div id="arrowSelectorLabel" class="ArrowSelector__label">Current</div>
  </div>
</template>

<script>
// Note to Salmon: Don't need to implement @keydown because <button> does it automatically!
// Huzzah to built-in functionalities!

export default {
  name: "ArrowSelector",
  props: {
    active: {
      type: String,
      required: true
    },
    items: {
      type: Array,
      required: true,
      validator: values =>
        values.every(
          value => typeof value === "string" || value instanceof String
        )
    }
  },
  computed: {
    current() {
      return this.items.indexOf(this.active) + 1;
    },
    total() {
      return this.items.length;
    }
  },
  methods: {
    setCurrent(next) {
      this.$emit("change", next);
    },
    handleLeftArrow() {
      this.setCurrent(this.current - 1 <= 0 ? this.total : this.current - 1);
    },
    handleRightArrow() {
      this.setCurrent(this.current + 1 > this.total ? 1 : this.current + 1);
    }
  },
  emits: ["change"]
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.ArrowSelector {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 80%;
}

.ArrowSelector__main {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-size: 3.6rem;
}

.ArrowSelector__btn {
  background-color: inherit;
  border: none;
  font-size: 7.2rem;
  cursor: pointer;
  color: @grey-05;
  transition: color 0.2s;

  &:focus,
  &:hover {
    color: @grey-04;
  }

  &:active {
    color: @grey-03;
  }
}

.ArrowSelector__label {
  font-size: 2.4rem;
  color: #4f4f4f;

  &::after {
    content: " Project";
  }
}

// iPhone 11 Pro Max, iPhone 8
@media only screen and (max-width: 414px) {
  .ArrowSelector__main {
    font-size: 3.6rem;
  }

  .ArrowSelector__btn {
    font-size: 4.8rem;
  }

  .ArrowSelector__label::after {
    content: "";
  }
}

// iPhone SE
@media only screen and (max-width: 320px) {
  .ArrowSelector__main {
    font-size: 2.4rem;
  }

  .ArrowSelector__btn {
    font-size: 4.8rem;
  }

  .ArrowSelector__label {
    font-size: 1.8rem;
  }
}
</style>
