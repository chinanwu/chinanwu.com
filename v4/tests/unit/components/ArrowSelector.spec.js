import { shallowMount } from "@vue/test-utils";

import ArrowSelector from "@/components/ArrowSelector.vue";

describe("ArrowSelector", () => {
  describe("rendering", () => {
    it("renders ArrowSelector", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          items: ["One", "Two"],
          active: "One"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);
      expect(wrapper.vm.total).toBe(2);
    });
  });

  describe("behaviour", () => {
    it("selects next if click right arrow", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          items: ["One", "Two"],
          active: "One"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);
      expect(wrapper.vm.total).toBe(2);

      wrapper.find("#arrowSelectorRight").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[2]]);
    });

    it("selects prev if click left arrow", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          items: ["One", "Two"],
          active: "Two"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(2);

      wrapper.find("#arrowSelectorLeft").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[1]]);
    });
  });

  describe("prop validation", () => {
    it("throws error if not valid items", () => {
      const validator = ArrowSelector.props.items.validator;
      expect(validator(["test"])).toBe(true);
      expect(validator([1])).toBe(false);
    });
  });
});
