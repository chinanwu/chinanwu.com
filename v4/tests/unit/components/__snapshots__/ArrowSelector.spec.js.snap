// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowSelector behaviour selects next if click right arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      aria-label="Select previous project"
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span
      aria-labelledby="arrowSelectorLabel"
    >
      1 / 2
    </span>
    <button
      aria-label="Select next project"
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
    id="arrowSelectorLabel"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects prev if click left arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      aria-label="Select previous project"
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span
      aria-labelledby="arrowSelectorLabel"
    >
      2 / 2
    </span>
    <button
      aria-label="Select next project"
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
    id="arrowSelectorLabel"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector rendering renders ArrowSelector 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      aria-label="Select previous project"
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span
      aria-labelledby="arrowSelectorLabel"
    >
      1 / 2
    </span>
    <button
      aria-label="Select next project"
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
    id="arrowSelectorLabel"
  >
    Current
  </div>
</div>
`;
